lib/AST/UnqualifiedLookup.cpp

```diff
@@ -401,19 +401,68 @@ bool implicitSelfReferenceIsUnwrapped(const ValueDecl *selfDecl,
   // and check that both its LHS and RHS are 'self'
   for (auto cond : conditionalStmt->getCond()) {
     if (auto pattern = cond.getPattern()) {
-      if (pattern->getBoundName() != Ctx.Id_self) {
+      bool isSelfRebinding = false;
+
+      if (pattern->getBoundName() == Ctx.Id_self) {
+        isSelfRebinding = true;
+      }
+
+      else if (auto OSP = dyn_cast<OptionalSomePattern>(pattern)) {
+        if (auto subPattern = OSP->getSubPattern()) {
+          if (subPattern->getBoundName() == Ctx.Id_self) {
+            isSelfRebinding = true;
+          }
+        }
+      }
+
+      if (!isSelfRebinding) {
         continue;
       }
     }
 
-    if (auto selfDRE = dyn_cast<DeclRefExpr>(cond.getInitializer())) {
-      return (selfDRE->getDecl()->getName().isSimpleName(Ctx.Id_self));
+    DeclRefExpr *condDRE = nullptr;
+    if (auto DRE = dyn_cast<DeclRefExpr>(cond.getInitializer())) {
+      condDRE = DRE;
+    }
+
+    if (auto LE = dyn_cast<LoadExpr>(cond.getInitializer())) {
```

Comment:
Hm, maybe we should use getSemanticsProvidingExpr here instead of dealing with LoadExpr directly? Should be able to simply all these checks into one

Reply:
We cannot lookup through LoadExpr but I think we still need to use getSemanticsProvidingExpr afterwards maybe in case decl ref was wrapped in parens or something? I’d structure the code in a way where it checks for load, gets semantic expr, checks for DeclRefExpr, WDYT?